File: cpp/bearing_area.cpp
/*
Copyright 2024 Lars Pastewka

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

#include <algorithm>
#include <iostream>

#include "bearing_area.h"


Eigen::ArrayXd nonuniform_bearing_area(Eigen::Ref<Eigen::ArrayXd> x, Eigen::Ref<Eigen::ArrayXd> h,
                                       Eigen::Ref<ArrayXl> el_sort_by_max, Eigen::Ref<Eigen::ArrayXd> heights) {
    if (x.size() != h.size()) {
        throw std::runtime_error("`x` and `h` must have the same size");
    }
    if (el_sort_by_max.size() != h.size()-1) {
        throw std::runtime_error("`el_sort_by_max` must have the same size one less than `h`");
    }

    /* The physical length of the line scan */
    const double physical_size{x.tail<1>().value() - x.head<1>().value()};

    /* Bearing area values for each input height */
    Eigen::ArrayXd fractional_bearing_areas(heights.size());

    /* Compute bearing areas */
    for (int j{0}; j < heights.size(); j++) {
        const double height{heights(j)};

        /* Find elements that cover the reference height */
        const auto lb{std::lower_bound(
            el_sort_by_max.begin(), el_sort_by_max.end(), height,
            [&h](const long& i, double value) {
                return std::max(h(i), h(i+1)) < value;
            })};

        /* Cumulative width for all elements completely below the input height */
        double bearing_area{0};

        for (auto k{lb}; k != el_sort_by_max.end(); k++) {
            const auto i{*k};
            const double hi{h(i)}, hi1{h(i+1)};
            const double dx{x(i+1) - x(i)};
            if (height < hi && height < hi1) {
                bearing_area += dx;
            } else if (!(height > hi && height > hi1)) {
                auto [minh, maxh] = std::minmax(hi, hi1);
                bearing_area += dx * (maxh - height) / (maxh - minh);
            }
        }
        fractional_bearing_areas(j) = bearing_area / physical_size;
    }

    return fractional_bearing_areas;
}


Eigen::ArrayXd uniform1d_bearing_area(Eigen::Ref<Eigen::ArrayXd> topography_h, bool periodic,
                                      Eigen::Ref<Eigen::ArrayXd> heights) {
    /* Bearing area values for each input height */
    Eigen::ArrayXd fractional_bearing_areas(heights.size());

    /* Compute bearing areas */
    const auto maxi{periodic ? topography_h.size() : topography_h.size()-1};
    for (int j{0}; j < heights.size(); j++) {
        double bearing_area{0};
        int physical_size{0};
        for (int i{0}; i < maxi; i++) {
            const auto i1{i < topography_h.size()-1 ? i+1 : 0};
            const double hi{topography_h(i)}, hi1{topography_h(i1)};
            /* Check for NaNs and only add if there are no NaNs */
            if (hi == hi && hi1 == hi1) {
                if (heights(j) < hi && heights(j) < hi1) {
                    bearing_area += 1;
                } else if (!(heights(j) > hi && heights(j) > hi1)) {
                    auto [minh, maxh] = std::minmax(hi, hi1);
                    bearing_area += (maxh - heights(j)) / (maxh - minh);
                }
                physical_size += 1;
            }
        }
        fractional_bearing_areas(j) = bearing_area / physical_size;
    }

    return fractional_bearing_areas;
}


double _triangle(double h1_in, double h2_in, double h3_in, double h) {
    double h1{h1_in}, h2{h2_in}, h3{h3_in};
    double bearing_area{0};

    /* Sort h1, h2, h3 in ascending order */
    if (h1 > h2)  std::swap(h1, h2);
    if (h2 > h3)  std::swap(h2, h3);
    if (h1 > h2)  std::swap(h1, h2);

    /* Compute bearing area */
    if (h <= h1) {
        bearing_area += 1.0;
    } else if (h1 < h && h <= h2) {
        bearing_area += 1 - (h - h1) * (h - h1) / ((h2 - h1) * (h3 - h1));
    } else if (h2 < h && h <= h3) {
        bearing_area += (h3 - h) * (h3 - h) / ((h3 - h2) * (h3 - h1));
    }
    return bearing_area;
}


Eigen::ArrayXd uniform2d_bearing_area(Eigen::Ref<RowMajorXXd> topography_h, bool periodic,
                                      Eigen::Ref<Eigen::ArrayXd> heights) {
    /* Number of grid points for looping */
    const auto nx{periodic ? topography_h.rows() : topography_h.rows()-1};
    const auto ny{periodic ? topography_h.cols() : topography_h.cols()-1};

    /* Bearing area values for each input height */
    Eigen::ArrayXd fractional_bearing_areas(heights.size());

    /* Compute bearing areas */
    for (int j{0}; j < heights.size(); j++) {
        double bearing_area{0};
        int projected_area{0};

        /* This is assuming column-major storage */
        for (int x{0}; x < nx; x++) {
            const auto x1{x < topography_h.rows()-1 ? x+1 : 0};
            for (int y{0}; y < ny; y++) {
                const auto y1{y < topography_h.cols()-1 ? y+1 : 0};
                const double h00{topography_h(x, y)};
                const double h10{topography_h(x1, y)};
                const double h01{topography_h(x, y1)};
                const double h11{topography_h(x1, y1)};
                /* Check for NaNs and only add triangle if there are no NaNs */
                if (h00 == h00 && h10 == h10 && h01 == h01) {
                    bearing_area += _triangle(h00, h10, h01, heights(j));
                    projected_area += 1;
                }
                /* Check for NaNs and only add triangle if there are no NaNs */
                if (h10 == h10 && h11 == h11 && h01 == h01) {
                    bearing_area += _triangle(h10, h11, h01, heights(j));
                    projected_area += 1;
                }
            }
        }
        fractional_bearing_areas(j) = bearing_area / projected_area;
    }

    return fractional_bearing_areas;
}

File: SurfaceTopography/Uniform/BearingArea.py
#
# Copyright 2024 Lars Pastewka
#
# ### MIT license
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
#

"""
Bearing area curve, also known as Abbott-Firestone curve or cumulative
distribution function of the surface heights.
"""

from functools import cached_property

import _SurfaceTopographyPP
import numpy as np

from ..HeightContainer import UniformTopographyInterface


class UniformBearingArea:
    def bounds(self, heights):
        """
        Compute bounds on the bearing area for a specific height. These bounds
        can be computed O(log(N)) time, where N is the number of elements in the
        line scan.

        Parameters
        ----------
        heights : float or np.ndarray
            Heights for which to compute the bearing area.

        Returns
        -------
        lower_bound : float or np.ndarray
            Lower bound on fractional area above a threshold height.
        upper_bound : float or np.ndarray
            Upper bound on fractional area above a threshold height.
        """
        el_min = np.searchsorted(self._el_min_heights, heights)
        el_max = np.searchsorted(self._el_max_heights, heights)
        return (self._nb_els - el_min) / self._nb_els, (self._nb_els - el_max) / self._nb_els

    @cached_property
    def min(self):
        return np.nanmin(self._h)  # There are NaNs if the original topography was masked

    @cached_property
    def max(self):
        return np.nanmax(self._h)  # There are NaNs if the original topography was masked


class Uniform1DBearingArea(UniformBearingArea):
    """
    Accelerated bearing area calculation for uniform line scans.
    """

    def __init__(self, h, is_periodic):
        self._is_periodic = is_periodic

        if self._is_periodic:
            self._el_min_heights = np.sort(np.ma.compressed(np.minimum(h, np.roll(h, -1))))
            self._el_max_heights = np.sort(np.ma.compressed(np.maximum(h, np.roll(h, -1))))
        else:
            self._el_min_heights = np.sort(np.ma.compressed(np.minimum(h[:-1], h[1:])))
            self._el_max_heights = np.sort(np.ma.compressed(np.maximum(h[:-1], h[1:])))
        self._nb_els = len(self._el_min_heights)  # This treats undefined data correctly
        self._h = np.ma.filled(h.astype(float), fill_value=np.nan)  # We pass the mask as NaNs to the C++ code

    def __call__(self, heights):
        """
        Compute the bearing area for a specific height.

        The bearing area as a function of height is also known as the
        Abbott-Firestone curve. If expressed as a fractional area, it is the
        cumulative distribution function of the surface heights. This function
        returns this fractional area.

        The function has complexity O(N) where N is the number of elements in


        Parameters
        ----------
        heights : float or np.ndarray
            Heights for which to compute the bearing area.

        Returns
        -------
        fractional_area : float or np.ndarray
            Fractional area above a the threshold height.
        """
        if np.isscalar(heights):
            return _SurfaceTopographyPP.uniform1d_bearing_area(self._h, self._is_periodic,
                                                               np.array([heights], dtype=float))[0]
        else:
            return _SurfaceTopographyPP.uniform1d_bearing_area(self._h, self._is_periodic,
                                                               heights.astype(float))


class Uniform2DBearingArea(UniformBearingArea):
    """
    Accelerated bearing area calculation for topographies.
    """

    def __init__(self, h, is_periodic):
        self._is_periodic = is_periodic

        if self._is_periodic:
            self._el_min_heights = np.sort(np.ma.compressed([
                np.minimum(np.minimum(h, np.roll(h, (-1, 0))), np.roll(h, (0, -1))),
                np.minimum(np.minimum(np.roll(h, (-1, -1)), np.roll(h, (-1, 0))), np.roll(h, (0, -1)))
            ]))
            self._el_max_heights = np.sort(np.ma.compressed([
                np.maximum(np.maximum(h, np.roll(h, (-1, 0))), np.roll(h, (0, -1))),
                np.maximum(np.maximum(np.roll(h, (-1, -1)), np.roll(h, (-1, 0))), np.roll(h, (0, -1)))
            ]))
        else:
            self._el_min_heights = np.sort(np.ma.compressed([
                np.minimum(np.minimum(h[:-1, :-1], h[1:, :-1]), h[:-1, 1:]),
                np.minimum(np.minimum(h[1:, 1:], h[1:, :-1]), h[:-1, 1:])
            ]))
            self._el_max_heights = np.sort(np.ma.compressed([
                np.maximum(np.maximum(h[:-1, :-1], h[1:, :-1]), h[:-1, 1:]),
                np.maximum(np.maximum(h[1:, 1:], h[1:, :-1]), h[:-1, 1:])
            ]))
        self._nb_els = len(self._el_min_heights)  # This treats undefined data correctly
        # We pass the mask as NaNs to the C++ code
        self._h = np.ascontiguousarray(np.ma.filled(h, fill_value=np.nan), dtype=float)

    def __call__(self, heights):
        """
        Compute the bearing area for a specific height.

        The bearing area as a function of height is also known as the
        Abbott-Firestone curve. If expressed as a fractional area, it is the
        cumulative distribution function of the surface heights. This function
        returns this fractional area.

        The function has complexity O(N) where N is the number of elements in


        Parameters
        ----------
        heights : float or np.ndarray
            Heights for which to compute the bearing area.

        Returns
        -------
        fractional_area : float or np.ndarray
            Fractional area above a the threshold height.
        """
        if np.isscalar(heights):
            return _SurfaceTopographyPP.uniform2d_bearing_area(self._h, self._is_periodic,
                                                               np.array([heights], dtype=float))[0]
        else:
            return _SurfaceTopographyPP.uniform2d_bearing_area(self._h, self._is_periodic,
                                                               heights.astype(float))


def bearing_area(self, heights=None):
    """
    Compute the bearing area for a specific height.

    The bearing area as a function of height is also known as the
    Abbott-Firestone curve. If expressed as a fractional area, it is the
    cumulative distribution function of the surface heights. This function
    returns this fractional area.

    Parameters
    ----------
    self : :obj:`UniformLineScan` or :obj:`Topography`
        Topography or line scan container object.
    heights : float or np.ndarray, optional
        Heights for which to compute the bearing area. (Default: None)

    Returns
    -------
    fractional_area : float or np.ndarray
        Fractional area above the threshold height, if height is given.
    bearing_area : :obj:`Uniform1DBearingArea` or :obj:`Uniform2DBearingArea`
        Instance of a class that caches the bearing area calculation.
    """
    if self.dim == 1:
        b = Uniform1DBearingArea(self.heights(), self.is_periodic)
    elif self.dim == 2:
        b = Uniform2DBearingArea(self.heights(), self.is_periodic)
    else:
        raise NotImplementedError('Bearing area is only implemented for 1D line scans and 2D topography maps.')

    if heights is None:
        return b
    else:
        return b(heights)


UniformTopographyInterface.register_function('bearing_area', bearing_area)
